### PR TITLE
tezos-encoding + derive: keep track of lifetime in NomReader

### PR DESCRIPTION
**Pull request by rafoo**
The `NomReader` trait is implicitly parameterized by the lifetime of the input byte slice. This commit makes the quantification on this explicit. This is needed to implement `NomReader` for structures which keep a reference to some slice of the input, for example to implement lazy deserialization.

### CODE REVIEW

File: tezos-encoding/src/lib.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@
 //!
 //! #[derive(Debug, PartialEq, HasEncoding, NomReader, BinWriter)]
 //! struct Outer<T>
-//! where T: Debug + PartialEq + HasEncoding + NomReader + BinWriter {
+//! where T: Debug + PartialEq + HasEncoding + (for <'a> NomReader<'a>) + BinWriter {
```
Comment (emturner) [inline tezos-encoding/src/lib.rs:32]:
```suggestion
//! where T: Debug + PartialEq + HasEncoding + for <'a> NomReader<'a> + BinWriter {
```

File: tezos-encoding/src/nom.rs [hunk 2]
```diff
@@ -572,6 +573,56 @@ where
     }
 }
 
+#[derive(Debug, PartialEq)]
+/// Lazy, memoized deserialization of a dynamic block.
+pub struct LazyCell<'a, T>
+where
+    T: NomReader<'a>,
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:580]: I don't think this requirement is correct - e.g. it will prevent using `LazyCell<Vec<u8>>`, for example.

By all means, `impl<T: NomReader> NomReader for LazyCell<T>` is correct, but requiring it on the struct itself is likely incorrect.

File: tezos-encoding/src/nom.rs [hunk 3]
```diff
@@ -572,6 +573,56 @@ where
     }
 }
 
+#[derive(Debug, PartialEq)]
+/// Lazy, memoized deserialization of a dynamic block.
+pub struct LazyCell<'a, T>
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:578]: this should probably exist in `types.rs` instead? (Like `Bytes`, `SizedBytes` etc)

Comment (emturner) [inline tezos-encoding/src/nom.rs:578]: Should this instead be an enum, to be equivalent to the `lazy` in `Data_encoding`?

```
Bytes(&'a [u8]), 
Value(T),
Both(&'a [u8], T)
```
This might allow slightly easier pattern matching as below

File: tezos-encoding/src/nom.rs [hunk 4]
```diff
@@ -572,6 +573,56 @@ where
...
+            deserialized: None,
+        })(input)
+    }
+}
+
+impl<'a, T> LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    pub fn to_bytes(self) -> &'a [u8] {
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:605]: not sure this should be in the public API? The bytes here won't necessarily correspond to just the bytes corresponding to the encoded T

File: tezos-encoding/src/nom.rs [hunk 5]
```diff
@@ -572,6 +573,56 @@ where
...
+        self.serialized
+    }
+
+    pub fn has_been_forced(&self) -> bool {
+        self.deserialized.is_some()
+    }
+
+    pub fn force<'b>(&'b mut self) -> Result<&'b T, nom::Err<NomError<'a>>> {
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:613]: why not just `fn value` instead of `force`?

File: tezos-encoding/src/nom.rs [hunk 6]
```diff
@@ -572,6 +573,56 @@ where
...
+
+impl<'a, T> LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    pub fn to_bytes(self) -> &'a [u8] {
+        self.serialized
+    }
+
+    pub fn has_been_forced(&self) -> bool {
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:609]: why is this needed?
If it's really needed, you could have a `fn value(&self) -> Option<&T>`?

File: tezos-encoding/src/nom.rs [hunk 7]
```diff
@@ -572,6 +573,56 @@ where
...
+    serialized: &'a [u8],
+    deserialized: Option<T>,
+}
+
+impl<'a, T> NomReader<'a> for LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    fn nom_read(input: &'a [u8]) -> NomResult<'a, Self> {
+        map(dynamic(rest), |serialized| LazyCell {
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:594]: does this need to require `dynamic` internally? It should delegate it perhaps to the underlying type?

(e.g. this will prevent using `LazyCell` with e.g. the  Rollup Inbox)

File: tezos-encoding/src/nom.rs [hunk 8]
```diff
@@ -572,6 +573,56 @@ where
...
+    T: NomReader<'a>,
+{
+    // Invariant: when the `deserialized` field is some, it's the
+    // result of deserialization of the `serialized` field using
+    // dynamic(T::nom_read)
+    serialized: &'a [u8],
+    deserialized: Option<T>,
+}
+
+impl<'a, T> NomReader<'a> for LazyCell<'a, T>
```
Comment (emturner) [inline tezos-encoding/src/nom.rs:589]: ideally would have a `impl From<T> for `LazyCell`?

and a `BinWriter` for `LazyCell` which internally would 'fail' the value if not present.

File: tezos-encoding-derive/src/nom.rs [hunk 9]
```diff
@@ -9,6 +9,7 @@ use crate::encoding::*;
 use proc_macro2::{Span, TokenStream};
```
Comment (emturner) [inline tezos-encoding-derive/src/nom.rs:9]: ideally, would have a `lazy` deriving possibility?

So this would enable things like:
```
pub struct Value {
  #[encoding(lazy, dynamic, bytes)]
  bytes1: LazyCell<Vec<u8>>,
  #[encoding(dynamic, lazy, bytes)]
  bytes2: LazyCell<Vec<u8>>
}
```

**Comment by emturner** [comment]
Thanks for the PR! Overall a big fan of adding something like this. 

Pretty happy with the first commit - most comments focused on the second (perhaps could be split out into a second PR if needed)

File: tezos-encoding/src/nom.rs (hunk 2 above)
Comment (rafoo) [reply tezos-encoding/src/nom.rs:580]: I wanted to prevent building a `LazyCell<'a, T>` for a `T` which does not implement `NomReader<'a>` because it would not be possible to `force` it.

File: tezos-encoding/src/nom.rs (hunk 3 above)
Comment (rafoo) [reply tezos-encoding/src/nom.rs:578]: I guess the difference is that Data_encoding not only features lazy deserialization but also lazy serialization. Do we need lazy serialization?

I'll try to see if turning the option into an enum simplifies the implementation.

File: tezos-encoding/src/nom.rs (hunk 4 above)
Comment (rafoo) [reply tezos-encoding/src/nom.rs:605]: It will never correspond, the length prefix from the `dynamic` encoding is missing here. After I move this to types.rs (as you suggested in https://github.com/trilitech/tezedge/pull/46#discussion_r1400585869) I will probably replace this by an `impl<'a> BinWriter for LazyCell<'a>`.

File: tezos-encoding/src/nom.rs (hunk 6 above)
Comment (rafoo) [reply tezos-encoding/src/nom.rs:609]: I don't know if we will ever need this but the corresponding feature in OCaml is needed in the protocol to carbonate lazy (de)serialization.

Let's remove it for now.

File: tezos-encoding/src/nom.rs (hunk 7 above)
Comment (rafoo) [reply tezos-encoding/src/nom.rs:594]: IIUC, you suggest to assume that `T` itself uses a `dynamic` encoding and to keep the size header in the `serialized` field. I don't know if we can enforce statically that `T` is a dynamic encoding.

File: tezos-encoding/src/nom.rs (hunk 3 above)
Comment (emturner) [reply tezos-encoding/src/nom.rs:578]: Ah good point; probably don't need `lazy_serialization` to be fair

**Review by emturner — APPROVED** [review]
